### PR TITLE
Normalize TFM platform versions when determining computed compatibility

### PR DESCRIPTION
**Pull request by advay26**
Addresses https://github.com/NuGet/NuGetGallery/issues/9711

### Problem:

TFMs that contain platform monikers and versions, like `net6.0-android31.0`, were not showing the `net7.0-android` and `net8.0-android` TFMs as users expected. 
https://www.nuget.org/packages/Xamarin.AndroidX.Lifecycle.Common/2.6.2.2#supportedframeworks-body-tab
image

If the TFM didn't have a platform version, say just `net6.0-android`, then the corresponding `net7.0` and `net8.0` TFMs show up as expected.
https://www.nuget.org/packages/Microsoft.CognitiveServices.Speech.Extension.ONNX.Runtime/1.33.0#supportedframeworks-body-tab
image

### Solution:

Currently, if a TFM (.NET 5+) has a platform version, it does not show any computed compatible TFMs. Our compatibility matrix only includes TFMs without platform versions.

To fix this, if we see a TFM with a platform version:
1. Create a normalized copy of the TFM, using an empty platform version 
    `net6.0-windows7.0` -> `net6.0-windows`
2. Use this normalized TFM to find computed compatible TFMs
    `net6.0-windows` -> `net6.0-windows`, `net7.0-windows`, `net8.0-windows`
3. Add the original TFM with its platform version (`net6.0-windows7.0`), and remove the normalized TFM (`net6.0-windows`) from the set
    `net6.0-windows7.0`, `net7.0-windows`, `net8.0-windows`

i.e. `net6.0-windows7.0` -> `net6.0-windows7.0`, `net7.0-windows`, `net8.0-windows`

Similarly,

| Asset TFM | Computed TFMs |
|-----------|---------------|
| `net6.0-windowsY.Z` | `net6.0-windowsY.Z`, `net7.0-windows`, `net8.0-windows` |
| `net6.0-windows` | `net6.0-windows`, `net7.0-windows`, `net8.0-windows` |
| `net6.0-windows7.0` | `net6.0-windows7.0`, `net7.0-windows`, `net8.0-windows` |
| `net6.0-windows99.0` | `net6.0-windows99.0`, `net7.0-windows`, `net8.0-windows` |
| `net7.0-windows7.0` | `net7.0-windows7.0`, `net8.0-windows` |
| `net6.0-ios` | `net6.0-ios`, `net7.0-ios`, `net8.0-ios` |
| `net6.0-ios15.0` | `net6.0-ios15.0`, `net7.0-ios`, `net8.0-ios` |

image

### CODE REVIEW

**Review by ryuyu — DISMISSED** [review]
So if my understanding is correct, this is change is made where it is because it would not be feasible for the CompatibilityMatrix to contain every possible iteration of a platform and version, right? :shipit:

**Comment by advay26** [comment]
> So if my understanding is correct, this is change is made where it is because it would not be feasible for the CompatibilityMatrix to contain every possible iteration of a platform and version, right? :shipit:

Yes, that's my understanding too. 

Also, the CompatibilityMatrix uses a list of 'supported frameworks' that we maintain ourselves https://github.com/NuGet/NuGetGallery/blob/main/src/NuGetGallery.Core/Frameworks/SupportedFrameworks.cs#L66, so we would need to manually update this list with every platform and platform version combo in order to cover it that way.

File: src/NuGetGallery.Core/Frameworks/FrameworkCompatibilityService.cs [hunk 1]
```diff
@@ -29,8 +30,26 @@ public static ISet<NuGetFramework> GetCompatibleFrameworks(IEnumerable<NuGetFram
                     continue;
                 }
 
-                if (CompatibilityMatrix.TryGetValue(packageFramework, out var compatibleFrameworks))
+                var normalizedPackageFramework = packageFramework;
+                bool hasPlatformVersion = false;
+
+                if ((packageFramework.Platform != string.Empty) && (packageFramework.PlatformVersion != FrameworkConstants.EmptyVersion))
```
Comment (erdembayar) [inline src/NuGetGallery.Core/Frameworks/FrameworkCompatibilityService.cs:36]: Not blocking. Could be more generic check `string.IsNotNullOrEmpty(packageFramework.Platform)` for 1st condition?

File: tests/NuGetGallery.Core.Facts/Frameworks/FrameworkCompatibilityServiceFacts.cs [hunk 2]
```diff
@@ -92,7 +92,15 @@ public void ValidPackageFrameworksReturnsFrameworksCompatibleForAtLeastOne(param
         }
 
         [Theory]
-        [InlineData("net6.0-windows7.0", "net6.0-windows", "net6.0-windows7.0", "net7.0-windows", "net7.0-windows7.0")]
+        [InlineData("net6.0-windows", "net6.0-windows", "net7.0-windows", "net8.0-windows")]
```
Comment (erdembayar) [inline tests/NuGetGallery.Core.Facts/Frameworks/FrameworkCompatibilityServiceFacts.cs:95]: Do we have test for simple case such as `net6.0` is compatible with `net5.0` etc?

**Review by erdembayar — COMMENTED** [review]
Just 2 questions.

File: src/NuGetGallery.Core/Frameworks/FrameworkCompatibilityService.cs (hunk 1 above)
Comment (ryuyu) [reply src/NuGetGallery.Core/Frameworks/FrameworkCompatibilityService.cs:36]: I think this is probably the more complete check since as written, I believe the condition is technically true if packageFramework.Platform is null (for whatever reason). Unless we WANT this to be true for null? @advay26 see https://learn.microsoft.com/en-us/dotnet/api/system.string?view=net-8.0#null-strings-and-empty-strings

Comment (advay26) [reply src/NuGetGallery.Core/Frameworks/FrameworkCompatibilityService.cs]: We want to treat null and empty platforms the same way here, so I've made the change

File: tests/NuGetGallery.Core.Facts/Frameworks/FrameworkCompatibilityServiceFacts.cs (hunk 2 above)
Comment (advay26) [reply tests/NuGetGallery.Core.Facts/Frameworks/FrameworkCompatibilityServiceFacts.cs:132]: I've added some general compatibility tests now too. I also added a new test case for `net5.0-windows`.

**Review by erdembayar — DISMISSED** [review]
👏 👏

**Review by ryuyu — DISMISSED** [review]

**Comment by joelverhagen** [comment]
This seems reasonable. Would it be possible to get a TFM compat expert on client to sign off on the concept here? It makes sense based on my understanding but I was not involved in the design or implementation of net5.0+ platform specific TFMs.

**Review by joelverhagen — DISMISSED** [review]

**Comment by nkolev92** [comment]
> If the TFM didn't have a platform version, say just net6.0-android, then the corresponding net7.0 and net8.0 TFMs show up as expected.

In  the .NET world with platforms, TargetPlatformVersion is mandatory. 
For the NuGetFramework type, no platform version means 0.0 (common pattern throughout), but it's restore and pack that prevent platform version-less packages/projects.

It is unusual that there are packages without a target platform in their package. In fact we have pack validation that should prevent that: https://learn.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu1012. 

Packages that contain assets like that were probably manually edited, and since restore tries not to be too opinionated and break it can work. 

I'm wondering if we should've blocked these packages from getting uploaded to nuget.org, but it's too late now I guess. 


Now focusing on the issue is that we want to show net8.0-android as a computed framework for net7.0-android31, I'm not really familiar with what versions are used, and how often people downgrade the default SDK platform version etc. 

In general, if the SDK defaults to net7.0-android31, for net8.0 it is likely the platform version is 32.0 or something like. 
If someone manually specifies net8.0-android30.0, then it'll be incompatible, but I don't think there'd be as much confusion as those people likely understand the platform version. 

The real concern is what is 31.0 wasn't the default, but a bumped version, and net8.0 defaults to a version 30.0, then this is a problem. 

I'd involve .NET folks here, as while I can speak on the technical correctness, they're the better authorities on what people would expect. 

Given net7.0-android31.0, I think showing net8.0-android31.0 is a good idea as a computed framework, which seems to what this PR is proposing.

**Comment by JonDouglas** [comment]
@nkolev92 i talked to daniel and immo on this. but while you point out these problems, the android team will never publish a TFM with a lesser value as a TPV. they build against the latest version of android APIs.

**Comment by JonDouglas** [comment]
also i think uno or some other groups do the manual creation b/c they have the net8.0-android TFM:

https://www.nuget.org/packages?q=&tfms=net8.0-android&prerel=true&sortby=relevance

**Comment by JonDouglas** [comment]
> I'm wondering if we should've blocked these packages from getting uploaded to nuget.org, but it's too late now I guess.

Should we track this issue to ensure people are using proper tooling that packs TPV?

**Comment by JonDouglas** [comment]
My only feedback here is to double check that the default major .NET version platform TPVs are right.

this: 
image

looks wrong to me initially. i think it is perfectly fine to compute and normalize them too if we don't have a good way of getting mappings. they can lookup docs or similar.

https://github.com/dotnet/docs/issues/38554

**Review by JonDouglas — DISMISSED** [review]
added comments for consideration

**Review by mariaghiondea — DISMISSED** [review]

**Comment by rolfbjarne** [comment]
Note that for iOS, we're planning on supporting multi-targeting, where both of these would be valid (at the same time):

* net8.0-ios17.0
* net8.0-ios17.1

> The Android platform version for each subsequent release of .NET will be >= to the previous .NET.

Same for iOS.

> I think for Android (and perhaps iOS is the same), you are going to have to know the explicitly valid combinations

That's not really feasible for iOS, we ship new versions with updated TPVs all year, depending on what Apple does. I'd really like to avoid a solution where someone would have to (remember to) update NuGet's knowledge all the time about which .NET versions supports which iOS versions.

We'll also change the default TPV even though we're not supposed to according to the spec, because the alternative is worse (customers' projects will stop compiling because Xcode was auto-updated by Apple, and they'd have to add the TPV manually to their project files to use the new Xcode, and we don't want people to use explicit TPV because then they'd have to update the TPV _again_ for the subsequent version of Xcode, and now the whole point of having a default value becomes useless because nobody can use it).

> We encourage users (via project templates and docs examples) to use the aliases that do not specify Android version.

I've seen more projects with the TPV than I'd expect, so I've wondered if the IDEs have been writing it out when saving csprojs without customers explicitly choosing it?

---

I think there are a few alternatives here though.

1. Given a `netX.0-iosY.Z`, then all versions of `net(X+1).0-ios` _should_ support `iOS Y.Z`, so you don't need the `Y.Z` for `X+1`.

    In other words, given `net6.0-ios15.4`, it can be expanded to list `net7.0-ios` and `net8.0-ios`.

2. The above solution isn't guaranteed (and I don't have a crystal ball), so a safer way to render it would be to somehow express that the version has a minimum:

    * `net6.0-ios15.4` ->
        * `net7.0-iosX` [where X is at least 15.4, but the iOS SDK may require it to be higher]
        * `net8.0-iosX` [where X is at least 15.4, but the iOS SDK may require it to be higher]

    This is more accurate, but certainly wordier, if not straight out uglier.

**Comment by JonDouglas** [comment]
@rolfbjarne 

> I've seen more projects with the TPV than I'd expect, so I've wondered if the IDEs have been writing it out when saving csprojs without customers explicitly choosing it?

Yeah I believe this was by design when packing w/ dotnet tooling.

What we will likely do in this case then @advay26 correct if wrong is that we will show:

`net6.0-android<TPV>` or `net6.0-android` and compute the normalized simplified compatible frameworks i.e. `net7.0-android` and `net8.0-android`. Similar story for iOS here. We will not add TPVs to computed compatible frameworks for now.

**Comment by advay26** [comment]
As @JonDouglas said, I've changed the proposal so that we're just showing the normalized compatible frameworks, rather than asserting a specific platform version. So a package targeting `net6.0-androidY.Z` will be mapped to `net6.0-androidY.Z`, `net7.0-android`, `net8.0-android`.

It would be difficult to continually keep track of all supported TPVs for each platform, so we'll instead try to point to docs somewhere once they're set up: https://github.com/dotnet/docs/issues/38554.

**Review by dannyjdev — APPROVED** [review]

**Review by ryuyu — APPROVED** [review]

**Review by mariaghiondea — APPROVED** [review]

**Review by erdembayar — APPROVED** [review]
👏 👏
